webservices/react-testing.md

```diff
@@ -0,0 +1,431 @@
+import { compose, flattenProp, withState } from "recompose";
+import addProps from "utilities/addProps";
+import { mealEvent } from "utilities/selectors";
+import srcMealDefault from "./mealDefault.svg";
+import styles from "./styles.module.scss";
+
+const mapStateToProps = (state, props) => ({
+  mealEvent: mealEvent(state, props)
+});
+
+export const setInitialImageUrl = ({ imageUrl }) => imageUrl;
+
+const dateTimeFormat = "h:mm A on MMM D, YYYY";
+export const dateTimeText = ({ consumedAt }) =>
+  `Log Time: ${moment(consumedAt).format(dateTimeFormat)}`;
+
+export const flagged = ({ flaggedAt }) => Boolean(flaggedAt);
+
+export const mealEventImgOnError = ({ setImageUrl }) => () => setImageUrl(null);
+
+export const defaultNoteText = "No Notes Available";
+export const noteText = ({ notes }) => (notes ? notes : defaultNoteText);
+
+export const srcMealEvent = ({ imageUrl }) =>
+  imageUrl ? imageUrl : srcMealDefault;
+
+export const stylesImage = ({ imageUrl, imageOrientation }) =>
+  imageUrl && imageOrientation === "Rotate 90 CW"
+    ? styles.imageMealEventRotated
+    : styles.imageMealEvent;
+
+export const styleNoteTextAbsent = "noteTextAbsent";
+export const styleNoteTextPresent = "noteTextPresent";
+export const stylesNoteText = ({ notes }) =>
+  notes ? styleNoteTextPresent : styleNoteTextAbsent;
+
+const timeFormat = "h:mm A";
+export const timeText = ({ consumedAt }) =>
+  moment(consumedAt).format(timeFormat);
+
+export default compose(
+  connect(mapStateToProps),
+  flattenProp("mealEvent"),
+  withState("imageUrl", "setImageUrl", setInitialImageUrl),
+  addProps({
+    dateTimeText,
+    flagged,
+    mealEventImgOnError,
+    noteText,
+    srcMealEvent,
+    stylesImage,
+    stylesNoteText,
+    timeText
+  })
+);
+```
+
+And here are the associated tests for it:
+
+```js
+import {
+  dateTimeText,
+  flagged,
+  defaultNoteText,
+  mealEventImgOnError,
+  noteText,
+  setInitialImageUrl,
+  srcMealEvent,
+  styleNoteTextAbsent,
+  styleNoteTextPresent,
+  styleNoteText,
+  stylesImage,
+  timeText,
+} from '../enhance';
+import srcMealDefault from '../mealDefault.svg';
+import styles from '../styles.module.scss';
+
+describe('dateTimeText', () => {
+  it('returns formatted full date with time', () => {
+    const result = dateTimeText({ consumedAt: '2017-05-08T16:00:19.433000Z' });
+
+    expect(result).toBe('Log Time: 12:00 PM on May 8, 2017');
+  });
+});
+
+describe('flagged', () => {
+  it('returns true if a date is given', () => {
+    const result = flagged({ flaggedAt: '2017-05-08T16:00:19.433000Z' });
+
+    expect(result).toBe(true);
+  });
+
+  it('returns false if a date is not given', () => {
+    const result = flagged({ flaggedAt: null });
+
+    expect(result).toBe(false);
+  });
+});
+
+describe('mealEventImgOnError', () => {
+  it('returns a function that calls setImageUrl with null', () => {
+    const setImageUrl = jest.fn();
+    const func = mealEventImgOnError({ setImageUrl });
+
+    func();
```

Comment:
why does `func` need to be defined and then immediately called instead of just calling `mealEventImgOnError({ setImageUrl })`?

Reply:
Calling `mealEventImgOnError` returns a function that calls `setImageUrl` with `null` when invoked. The returned function needs to be invoked for the expectation to pass.